site/_posts/2023-02-03-single-sign-on-laravel-fusionauth.md

```diff
@@ -0,0 +1,388 @@
+
+```shell
+$ ./vendor/bin/sail artisan make:migration add_fusionauth_fields_user_table
+```
+
+Then, go to your `database/migrations` folder, where you'll find a file there with the name you created and prefixed by
+today's date (e.g `2023_02_03_123000_add_fusionauth_fields_user_table.php`).
+
+Let's edit that file so we can describe the changes needed in our database. We are going to add three new columns to
+the `users` table, `fusionauth_id`, `fusionauth_token` and `fusionauth_refresh_token`, and change the `password` column
+so it can be `NULL`.
+
+```php
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+return new class extends Migration
+{
+    /**
+     * When running this migration, we want to create three new columns and change the password so it can be NULL
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('users',  function (Blueprint $table) {
+            $table->string('fusionauth_id', 36)->unique();
+            $table->text('fusionauth_token');
+            $table->text('fusionauth_refresh_token')->nullable();
+            $table->string('password')->nullable()->change();
+        });
+    }
+
+    /**
+     * When rolling back this migration, we want to drop the added columns and revert the password to NOT NULL again
+     *
+     * @return void
+     */
+    public function down()
+    {
+        Schema::table('users',  function (Blueprint $table) {
+            $table->dropColumn('fusionauth_id');
+            $table->dropColumn('fusionauth_token');
+            $table->dropColumn('fusionauth_refresh_token');
+            $table->string('password')->nullable(false)->change();
+        });
+    }
+};
+```
+
+We also need to add `doctrine/dbal` as a dependency, which is a package responsible for changing table and column
+definitions:
+
+```shell
+$ ./vendor/bin/sail composer require doctrine/dbal
+```
+
+Let's run our migration to add those new columns:
+
+```shell
+$ ./vendor/bin/sail php artisan migrate
+
+   INFO  Running migrations.  
+
+  2023_02_03_123000_add_fusionauth_fields_user_table ........................... 51ms DONE
+```
+
+Now, let's change our User model so it can become aware of these new columns. Modify `app/Models/User.php` and add them
+to both `$fillable` and `$hidden` properties:
+
+```php
+/**
+ * The attributes that are mass assignable.
+ *
+ * @var array<int, string>
+ */
+protected $fillable = [
+    'name',
+    'email',
+    'password',
+    'fusionauth_id',            // Add this column here
+    'fusionauth_token',         // Add this column here
+    'fusionauth_refresh_token', // Add this column here
+];
+
+/**
+ * The attributes that should be hidden for serialization.
+ *
+ * @var array<int, string>
+ */
+protected $hidden = [
+    'password',
+    'remember_token',
+    'fusionauth_id',            // Add this column here
```

Comment:
Why hide this? This is an immutable attribute of the user now.

Reply:
[Laravel's Eloquent ORM](https://laravel.com/docs/9.x/eloquent-serialization#hiding-attributes-from-json) uses this attribute to hide data when serializing the model. They could have chosen a better name though...